lib/spack/spack/parser.py

```diff
@@ -0,0 +1,458 @@
+# Copyright 2013-2022 Lawrence Livermore National Security, LLC and other
+# Spack Project Developers. See the top-level COPYRIGHT file for details.
+#
+# SPDX-License-Identifier: (Apache-2.0 OR MIT)
+"""Parser for spec literals"""
+import pathlib
+import re
+from enum import Enum, auto
+from typing import Iterator, List, Optional
+
+from llnl.util.tty import color
+
+import spack.error
+import spack.spec
+import spack.variant
+import spack.version
+
+#: Valid name for specs and variants. Here we are not using
+#: the previous "w[\w.-]*" since that would match most
+#: characters that can be part of a word in any language
+VALID_NAME_WITHOUT_DOTS = r"[a-zA-Z_0-9][a-zA-Z_0-9\-]*"
+VALID_NAME_WITH_DOTS = rf"{VALID_NAME_WITHOUT_DOTS}(\.{VALID_NAME_WITHOUT_DOTS})+"
+VALID_GIT_VERSION = rf"(((git.)?({VALID_NAME_WITH_DOTS}))|({VALID_NAME_WITHOUT_DOTS}))"
+
+VALID_NAME = r"[a-zA-Z_0-9][a-zA-Z_0-9\-.]*"
+
+VALID_HASH = r"[a-zA-Z_0-9]+"
+
+#: A filename starts either with a "." or a "/" or a "{name}/"
+VALID_FILENAME = r"(\.|\/|[a-zA-Z0-9-_]*\/)([a-zA-Z0-9-_\.\/]*)(\.json|\.yaml)"
+
+VALID_VALUE = r"([a-zA-Z_0-9\-+\*.,:=\~\/\\]+)"
+VALID_QUOTED_VALUE = r"[\"']+([a-zA-Z_0-9\-+\*.,:=\~\/\\\s]+)[\"']+"
+
+VALID_VERSION = r"([a-zA-Z_0-9-.][a-zA-Z_0-9-.]*)"
+VALID_VERSION_RANGE = rf"({VALID_VERSION}:{VALID_VERSION}|:{VALID_VERSION}|{VALID_VERSION}:|:)"
+VALID_VERSION_LIST = (
+    rf"({VALID_VERSION_RANGE}|{VALID_VERSION})([,]({VALID_VERSION_RANGE}|{VALID_VERSION}))*"
+)
+
+
+class TokenKind(Enum):
+    """Enumeration of the different token kinds in the spec grammar"""
+
+    # FILENAME
+    FILENAME = auto()
+    # DAG hash
+    DAG_HASH = auto()
+    # Dependency
+    DEPENDENCY = auto()
+    # Version
+    VERSION_HASH_PAIR = auto()
+    VERSION = auto()
+    # Variants
+    PROPAGATED_BOOL_VARIANT = auto()
+    BOOL_VARIANT = auto()
+    KEY_VALUE_PAIR = auto()
+    PROPAGATED_KEY_VALUE_PAIR = auto()
+    # Compilers
+    COMPILER_AND_VERSION = auto()
+    COMPILER = auto()
+    # Package name
+    UNQUALIFIED_PACKAGE_NAME = auto()
+    FULLY_QUALIFIED_PACKAGE_NAME = auto()
+    # White spaces
+    WS = auto()
+    # Unexpected character
+    UNEXPECTED = auto()
```

Comment:
Here I did everything except:

> [ ... ] for tokens like UNEXPECTED that aren't associated with a particular parse, you can just indicate that with None.

The idea of that token is that it can be extended later, i.e. we have a valid set of regexes and we might add common error patterns afterwards - in following PRs. When we match an error pattern we can emit a more meaningful message, like "The `@` sigil cannot be repeated multiple times" and point to the actual parsing error, since each token has a `start` and an `end` position in the text stream.